keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+##### Integration tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- <test>: <link to test coverage>
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+- <test>: <link to test coverage>
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete benchmarks
+
+#### GA
+
+- Add metrics
+- Add a full documentation with examples for the most popular providers, e.g., Okta, Dex, Auth0
+- Migration guide
+- Deprecation warnings for non-stuctured OIDC provider configuration
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+kube-apiserver `--oidc-*` flags require deprecation warnings on the stable release of the feature.
+It is possible to react only to the `--oidc-issuer-url` flag because other flags cannot be enabled separately from this one.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- Feature gate
+  - Feature gate name: `StructuredOIDCConfiguration`
+  - Components depending on the feature gate:
+    - kube-apiserver
+
+```go
+FeatureSpec{
+	Default: false,
+	LockToDefault: false,
+	PreRelease: featuregate.Alpha,
+}
+```
+
+###### Does enabling the feature change any default behavior?
+
+No.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, but disable means also deleting the flag from the kube-apiserver manifest.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+No impact.
+
+###### Are there any tests for feature enablement/disablement?
+
+Not required.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+It cannot fail.
+
+###### What specific metrics should inform a rollback?
+
+No specific metrics are required.
```

Comment:
Would any of the existing kube-apiserver metrics around authentication point to a potential misconfiguration of the new structured config or a bug in parsing the config?

Reply:
I think we can add some metrics for the cases when the authenticator uses insecure configuration, e.g. if it skips basic OIDC validations.

Reply:
If you're considering hot-reloads of the config, it would make sense to expose config read failures in a metric.